lib/msf/core/db_manager/note/initialize_notes.rb

```diff
@@ -0,0 +1,120 @@
+          sname = nil
+          proto_lower = opts[:proto].to_s.downcase # Catch incorrect usages
+          case proto_lower
+          when 'tcp','udp'
+            proto = proto_lower
+            sname = opts[:sname] if opts[:sname]
+            # XXX: These normalizations are lazy af
+          when 'http', 'smb'
+            proto = 'tcp'
+            sname = proto_lower
+          when 'dns','snmp','dhcp'
+            proto = 'udp'
+            sname = proto_lower
+          else
+            proto = 'tcp'
+            sname = proto_lower
+          end
+          sopts = {
+            :workspace => wspace,
+            :host  => host,
+            :port  => opts[:port],
+            :proto => proto
+          }
+          sopts[:name] = sname if sname
+          report_service(sopts)
+        end
+      end
+      # Update Modes can be :unique, :unique_data, :insert
+      mode = opts[:update] || :unique
+
+
+      if addr and not host
+        host = get_host(:workspace => wspace, :host => addr)
+      end
+      if host and (opts[:port] and proto)
+        # only one result can be returned, as the +port+ field restricts potential results to a single service
+        service = services(:workspace => wspace,
+                           :hosts => {address: host.address},
+                           :proto => proto,
+                           :port => opts[:port]).first
+      elsif opts[:service] and opts[:service].kind_of? ::Mdm::Service
+        service = opts[:service]
+      end
+
+      ntype  = opts.delete(:type) || opts.delete(:ntype) || (raise RuntimeError, "A note :type or :ntype is required")
+      data   = opts[:data]
+      note   = nil
+      conditions = { :ntype => ntype }
+      conditions[:host_id] = host[:id] if host
+      conditions[:service_id] = service[:id] if service
+      conditions[:vuln_id] = opts[:vuln_id]
+
+      case mode.to_sym
+      when :unique
+        note      = wspace.notes.where(conditions).first_or_initialize
+        note.data = data
+      when :unique_data
+        notes = wspace.notes.where(conditions)
+
+        # Don't make a new Note with the same data as one that already
+        # exists for the given: type and (host or service)
+        notes.each do |n|
+          # Compare the deserialized data from the table to the raw
+          # data we're looking for.  Because of the serialization we
+          # can't do this easily or reliably in SQL.
+          if n.data == data
+            note = n
+            break
+          end
+        end
+        if not note
+          # We didn't find one with the data we're looking for, make
+          # a new one.
+          note = wspace.notes.new(conditions.merge(:data => data))
+        end
+      else
+        # Otherwise, assume :insert, which means always make a new one
+        note = wspace.notes.new
+        if host
+          note.host_id = host[:id]
+        end
+        if opts[:service] and opts[:service].kind_of? ::Mdm::Service
```

Comment:
```suggestion
        if opts[:service] && opts[:service].kind_of? ::Mdm::Service
```

Reply:
This is weird, I changed it over but it's giving me a syntax error and I don't understand why
it works if I use `and` or if I do `&& opts[:service].kind_of?(::Mdm::Service)` and I can't think of any reason why that would be

![image](https://user-images.githubusercontent.com/19910435/142430567-c6d7edc2-2265-40e8-bff1-7e8361dc6a67.png)

Reply:
I suspect the parsing is done differently based on `&&` or `and`. `opts[:service].kind_of?(::Mdm::Service)` seems reasonable.